### ISSUE TITLE
Modified Conv works only once

### ISSUE THREAD

**Issue by SaharMagdyy** [ultralytics/ultralytics#21285]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

I’ve been working with and modifying YOLO models since YOLOv5, and I’ve recently encountered a new and unexpected issue with YOLOv8 that I haven't seen before.

Problem Description: 

1) The steps I followed:
I cloned the official Ultralytics repository and installed it in editable mode using the following steps:

!git clone https://github.com/ultralytics/ultralytics

%cd /kaggle/working/ultralytics

!pip install -e .

I then modified the file ultralytics/nn/modules/conv.py. Specifically, I added a simple print statement inside the Conv class to test that my changes were being picked up:

For example:

class Conv(nn.Module):
    """
    Standard convolution module with batch normalization and activation.

    Attributes:
        conv (nn.Conv2d): Convolutional layer.
        bn (nn.BatchNorm2d): Batch normalization layer.
        act (nn.Module): Activation function layer.
        default_act (nn.Module): Default activation function (SiLU).
    """

    default_act = nn.SiLU()  # default activation

    def __init__(self, c1, c2, k=1, s=1, p=None, g=1, d=1, act=True):
        """
        Initialize Conv layer with given parameters.

        Args:
            c1 (int): Number of input channels.
            c2 (int): Number of output channels.
            k (int): Kernel size.
            s (int): Stride.
            p (int, optional): Padding.
            g (int): Groups.
            d (int): Dilation.
            act (bool | nn.Module): Activation function.
        """
        super().__init__()
        self.conv = nn.Conv2d(c1, c2, k, s, autopad(k, p, d), groups=g, dilation=d, bias=False)
        self.bn = nn.BatchNorm2d(c2)
        self.act = self.default_act if act is True else act if isinstance(act, nn.Module) else nn.Identity()

    def forward(self, x):
        """
        Apply convolution, batch normalization and activation to input tensor.

        Args:
            x (torch.Tensor): Input tensor.

        Returns:
            (torch.Tensor): Output tensor.
        """
       print("hello")
        return self.act(self.bn(self.conv(x)))

    def forward_fuse(self, x):
        """
        Apply convolution and activation without batch normalization.

        Args:
            x (torch.Tensor): Input tensor.

        Returns:
            (torch.Tensor): Output tensor.
        """
        print("hello")
        return self.act(self.conv(x))

To verify the changes, I ran:

from ultralytics import YOLO

# Load a model
model = YOLO('yolov8n.yaml')  # build a new model from scratch
model = YOLO('yolov8n.pt')  # load a pretrained model (recommended for training)

The output correctly printed multiple "hello" messages as expected.

**The Issue**
However, the problem arises when I remove or modify the print("hello") lines (or make any other change in conv.py). No matter what I do afterward, the output stays the same, as if the model is still running the initial version with the original modifications.

The same happens in reverse: if I run the code without modification first, then apply changes later, those changes are never reflected. It’s as if YOLO is caching or freezing the state of the code at first execution and not reloading the modified files on subsequent runs.

**Additional Info**

1. I'm working within a Kaggle notebook environment.
2. I’ve confirmed that I’m importing the correct local module.
3. I’ve used pip install -e . to ensure editable mode is active.

**Request**
Could you please clarify:
Is there any internal caching or model tracing mechanism in YOLOv8 that might prevent the framework from seeing recent code edits?

Is there a recommended way to force YOLO to reload local code after modification?

Any guidance or clarification would be highly appreciated. Thank you for your great work on the project.

### Additional

I have seen issue #21160 but didnt know what the solution is still..

**Comment by UltralyticsAssistant** [comment]
👋 Hello @SaharMagdyy, thank you for your detailed report and for your interest in Ultralytics 🚀! We recommend visiting the [Docs](https://docs.ultralytics.com/) for new users, where you can find many helpful [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples. Many common questions may already be answered there.

As this is a 🐛 Bug Report related to code modifications, please ensure you provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) (MRE) demonstrating the issue, including the steps you follow after modifying the code, so our team can best assist you.

If you have questions about custom training or environment setup, please provide as much detail as possible, including environment specifics, dataset image examples, and logs, and verify you’re following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best! For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package and ensure all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) are installed in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO can be run in these up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

---

This is an automated response. An Ultralytics engineer will also review your issue and provide further assistance soon. Thank you for helping us improve YOLO! 🦾

**Comment by Y-T-G** [comment]
This is just how Python works. Modifications will only take effect after you restart the runtime/kernel.

You can use `autoreload` to make the reloading automatic:

https://stackoverflow.com/questions/1907993/autoreload-of-modules-in-ipython

It doesn't always work though. So sometimes you still need to restart the kernel.
